astropy/table/table.py

```diff
@@ -1638,6 +1716,19 @@ def add_columns(self, cols, indexes=None, copy=True, rename_duplicate=False):
                 new_indexes.insert(i, None)
                 newcols.insert(i, col)
 
+        if names is None:
+            names = ['col{}'.format(name) for name in range(len(self.columns), len(self.columns) + len(cols))]
+            for col, name in zip(cols, names):
+                if col.info.name is None:
+                    col.info.name = name
```

Comment:
Here you'd need to check also that the column is not part of another table.

But I wonder if one cannot combine the two loops. E.g.,
```
if names is None:
    names = (None,) * len(cols)
elif len(names) != len(cols):
    raise ValueError('Number of names must match number of columns')

for i, (col, name) in enumerate(zip(cols, names)):
    if name is None:
        if col.info.name is not None:
            continue
        name = 'col{}'.format(i + len(self.columns))
    if col.info.parent_table is not None:
        col = col_copy(col)
    col.info.name = name
```